plugins/terraform/resource_environment_test.go

```diff
@@ -0,0 +1,140 @@
+package main
+
+import (
+	"github.com/golang/mock/gomock"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/quintilesims/layer0/common/models"
+	"testing"
+)
+
+func TestEnvironmentCreate_basic(t *testing.T) {
+	ctrl, mockClient, provider := setupUnitTest(t)
+	defer ctrl.Finish()
+
+	mockClient.EXPECT().
+		CreateEnvironment("test-env", "m3.medium", 0, []byte("")).
+		Return(&models.Environment{EnvironmentID: "eid"}, nil)
+
+	mockClient.EXPECT().
+		GetEnvironment("eid").
+		Return(&models.Environment{}, nil)
+
+	environmentResource := provider.ResourcesMap["layer0_environment"]
+	d := schema.TestResourceDataRaw(t, environmentResource.Schema, map[string]interface{}{
+		"name": "test-env",
+	})
+
+	if err := environmentResource.Create(d, mockClient); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestEnvironmentCreate_specifyOptional(t *testing.T) {
+	ctrl, mockClient, provider := setupUnitTest(t)
+	defer ctrl.Finish()
+
+	mockClient.EXPECT().
+		CreateEnvironment("test-env", "m3.large", 2, []byte("user data")).
+		Return(&models.Environment{EnvironmentID: "eid"}, nil)
+
+	mockClient.EXPECT().
+		GetEnvironment("eid").
+		Return(&models.Environment{}, nil)
+
+	environmentResource := provider.ResourcesMap["layer0_environment"]
+	d := schema.TestResourceDataRaw(t, environmentResource.Schema, map[string]interface{}{
+		"name":      "test-env",
+		"size":      "m3.large",
+		"min_count": 2,
+		"user_data": "user data",
+	})
+
+	if err := environmentResource.Create(d, mockClient); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestEnvironmentRead(t *testing.T) {
+	ctrl, mockClient, provider := setupUnitTest(t)
+	defer ctrl.Finish()
+
+	mockClient.EXPECT().
+		GetEnvironment("eid").
+		Return(&models.Environment{}, nil)
+
+	environmentResource := provider.ResourcesMap["layer0_environment"]
+	d := schema.TestResourceDataRaw(t, environmentResource.Schema, map[string]interface{}{})
+	d.SetId("eid")
+
+	if err := environmentResource.Read(d, mockClient); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestEnvironmentUpdate(t *testing.T) {
+	ctrl, mockClient, provider := setupUnitTest(t)
+	defer ctrl.Finish()
+
+	gomock.InOrder(
+		mockClient.EXPECT().
+			CreateEnvironment("test-env", "m3.medium", 0, []byte("")).
+			Return(&models.Environment{EnvironmentID: "eid"}, nil),
+		// Return(&createdEnvironment, nil),
+
+		mockClient.EXPECT().
+			GetEnvironment("eid").
+			Return(&models.Environment{EnvironmentID: "eid"}, nil),
+		// Return(&createdEnvironment, nil),
+
+		mockClient.EXPECT().
+			UpdateEnvironment("eid", 3).
+			Return(&models.Environment{EnvironmentID: "eid"}, nil),
+		// Return(&updatedEnvironment, nil),
+
+		mockClient.EXPECT().
+			GetEnvironment("eid").
+			Return(&models.Environment{EnvironmentID: "eid"}, nil),
+		// Return(&updatedEnvironment, nil),
+	)
+
+	environmentResource := provider.ResourcesMap["layer0_environment"]
+	d1 := schema.TestResourceDataRaw(t, environmentResource.Schema, map[string]interface{}{
```

Comment:
Yeah, I was having some issues finding the best way to trigger a change.

At one point, I had something like this:

```
	environmentResource := provider.ResourcesMap["layer0_environment"]
	d := schema.TestResourceDataRaw(t, environmentResource.Schema, map[string]interface{}{
		"name": "test-env",
	})

	if err := environmentResource.Create(d1, mockClient); err != nil {
		t.Fatal(err)
	}

	fmt.Printf("has change: %t\n", d1.HasChange("min_count"))
	// false, as expected

	d.Set("min_count", 3)
	d.SetId("eid")

	fmt.Printf("has change: %t\n", d1.HasChange("min_count"))
	// false, which is NOT expected

	if err := environmentResource.Update(d, mockClient); err != nil {
		t.Fatal(err)
	}
```

It seemed to me that setting up two different `TestResourceDataRaw` was the way to trigger `HasChange` updates, but it's entirely possible that I've overlooked or just not found the actual right way to do it.